### PR TITLE
MapValue clean up

### PR DESCRIPTION
**Pull request by pontusmelke**
We are in many places leaking the internal implementation of MapValue which makes it hard to try out different implementations. This PR removes all leakages of the internal `java.util.Map` of `MapValue`

### CODE REVIEW

File: community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/BoltRequestMessageReader.java [hunk 1]
```diff
@@ -101,8 +100,10 @@ public void read( BoltRequestMessageHandler handler ) throws IOException
...
         MapValue authTokenValue = unpacker.unpackMap();
         AuthTokenValuesWriter writer = new AuthTokenValuesWriter();
-        return authTokenValue.entrySet()
-                .stream()
-                .collect( toMap( Map.Entry::getKey, entry -> writer.valueAsObject( entry.getValue() ) ) );
+        HashMap<String,Object> tokenMap = new HashMap<>( authTokenValue.size() );
+
+        authTokenValue.foreach(
+                ( key, value ) -> tokenMap.put( key, writer.valueAsObject( value ) ) );
+        return tokenMap;
```
Comment (SaschaPeukert) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/BoltRequestMessageReader.java:107]: This is way more readable! Thanks!

File: community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/commands/expressions/DesugaredMapProjection.scala [hunk 2]
```diff
@@ -40,12 +37,18 @@ case class DesugaredMapProjection(id: String, includeAllProps: Boolean, literalE
       case v if v == Values.NO_VALUE => return Values.NO_VALUE
       case IsMap(m) => if (includeAllProps) m(state.query) else VirtualValues.emptyMap()
     }
-    val mapOfLiteralValues = literalExpressions.map {
-      case (k, e) => (k, e(ctx, state))
-    }.toMap.asJava
+    val builder = new MapValueBuilder()
```
Comment (SaschaPeukert) [inline community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/commands/expressions/DesugaredMapProjection.scala]: ```val builder = new MapValueBuilder(literalExpressions.size)```  ?

File: community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java [hunk 3]
```diff
@@ -106,9 +105,9 @@ public AnyValue mapRelationship( VirtualRelationshipValue value )
         @Override
         public AnyValue mapMap( MapValue value )
         {
-            Map<String,AnyValue> map = new HashMap<>();
-            value.foreach( ( k, v ) -> map.put( k, v.map( this ) ) );
-            return VirtualValues.map( map );
+            MapValueBuilder builder = new MapValueBuilder();
```
Comment (SaschaPeukert) [inline community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java:108]: we could initialize this new ```MapValueBuilder``` with ```value.size```, right?

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/CypherAdapterStream.java [hunk 4]
```diff
@@ -152,24 +150,26 @@ public static AnyValue convert( Iterable<Notification> notifications )
             List<AnyValue> out = new ArrayList<>();
             for ( Notification notification : notifications )
             {
-                Map<String,AnyValue> notificationMap = new HashMap<>( 4 );
-                notificationMap.put( "code", stringValue( notification.getCode() ) );
-                notificationMap.put( "title", stringValue( notification.getTitle() ) );
-                notificationMap.put( "description", stringValue( notification.getDescription() ) );
-                notificationMap.put( "severity", stringValue( notification.getSeverity().toString() ) );
+                MapValueBuilder builder = new MapValueBuilder();
```
Comment (SaschaPeukert) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/CypherAdapterStream.java]: ```new MapValueBuilder(4);```

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/ExecutionPlanConverter.java [hunk 5]
```diff
@@ -44,21 +43,21 @@ private ExecutionPlanConverter()
 
     public static MapValue convert( ExecutionPlanDescription plan )
     {
-        Map<String,AnyValue> out = new HashMap<>();
-        out.put( "operatorType", stringValue( plan.getName() ) );
-        out.put( "args", ValueUtils.asMapValue( plan.getArguments() ) );
-        out.put( "identifiers", ValueUtils.asListValue( plan.getIdentifiers() ) );
-        out.put( "children", children( plan ) );
+        MapValueBuilder out = new MapValueBuilder(  );
```
Comment (SaschaPeukert) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/ExecutionPlanConverter.java]: Could it make sense to initialize this one with a size?

**Review by SaschaPeukert — COMMENTED** [review]
Awesome work, just a few minor comments and you might need to rebase it

File: community/values/src/main/java/org/neo4j/values/virtual/MapValueBuilder.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.values.virtual;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.neo4j.values.AnyValue;
+
+public class MapValueBuilder
```
Comment (fickludd) [inline community/values/src/main/java/org/neo4j/values/virtual/MapValueBuilder.java:27]: Wow, using a builder pattern here is great. Awesome.

File: community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/MapSupport.scala [hunk 7]
```diff
@@ -84,7 +85,10 @@ class LazyMap[T](ctx: QueryContext, ops: Operations[T], id: Long)
   override def size(): Int = allProps.size()
 
   //we need a way forcefully load lazy values
-  def load(): Boolean = allProps != null
+  def load(): MapValue =
```
Comment (fickludd) [inline community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/MapSupport.scala]: If we really need lazy values, then maybe we should upgrade `load()` to something like `MapValue.materialize()` so all maps had this functionality. In general I'm afraid of any lazy values and the bugs they might cause. This also applies to lazily filtered and mapped values, where the filtering/mapping function might not give the same result at the time of execution when it's lazy.

**Review by fickludd — COMMENTED** [review]
I have only skimmed this, but there are some nice improvements here. In particular like the `MapValueBuilder`. 

I think when the `ListValue` implementations were introduced: There should be some test to assert that all the different `MapValue` implementations are equal (and share hashcode) if they have the same contents, and that they are not equal when they have different contents.

File: community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java (hunk 3 above)
Comment (pontusmelke) [reply community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java:108]: `value.size()` may be a O(n) operation so I think I prefer not to do it in this case.

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/CypherAdapterStream.java (hunk 4 above)
Comment (pontusmelke) [reply community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/CypherAdapterStream.java]: I'm torn in this case it is mostly four but in some cases if `includePosition` is true we include a fifth element, so maybe go with size 5 instead.

**Review by fickludd — APPROVED** [review]

**Pull request merged**